modules/face/src/regtree.cpp

```diff
@@ -0,0 +1,363 @@
+                for(unsigned long k=0;k<(*samples)[j].shapeResiduals.size();k++){
+                    (*sum)[k]=(*sum)[k]+(*samples)[j].shapeResiduals[k];
+                }
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+        vector<Point2f>* sum;
+};
+class modifySamples : public ParallelLoopBody
+{
+    public:
+        modifySamples(vector<training_sample>* samples_,vector<Point2f>* temp_,float* learning_rate_) :
+        samples(samples_),
+        temp(temp_),
+        learning_rate(learning_rate_)
+        {
+        }
+        virtual void operator()( const cv::Range& range) const
+        {
+            for (size_t j = range.start; j < (size_t)range.end; ++j){
+                for(unsigned long k=0;k<(*samples)[j].shapeResiduals.size();k++){
+                    (*samples)[j].shapeResiduals[k]=(*samples)[j].shapeResiduals[k]-(*learning_rate)*(*temp)[k];
+                    (*samples)[j].current_shape[k]=(*samples)[j].actual_shape[k]-(*samples)[j].shapeResiduals[k];
+                }
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+        vector<Point2f>* temp;
+        float* learning_rate;
+};
+class splitSamples : public ParallelLoopBody
+{
+    public:
+        splitSamples(vector<training_sample>* samples_,vector< vector<Point2f> >* leftsumresiduals_,vector<unsigned long>* left_count_,unsigned long* num_test_splits_,vector<splitr>* feats_) :
+        samples(samples_),
+        leftsumresiduals(leftsumresiduals_),
+        left_count(left_count_),
+        num_test_splits(num_test_splits_),
+        feats(feats_)
+        {
+        }
+        virtual void operator()( const cv::Range& range) const
+        {
+            for (size_t i = range.start; i <(size_t) range.end; ++i){
+                for(unsigned long j=0;j<*(num_test_splits);j++){
+                    (*left_count)[j]++;
+                    if ((float)(*samples)[i].pixel_intensities[(*feats)[j].index1] - (float)(*samples)[i].pixel_intensities[(*feats)[j].index2] > (*feats)[j].thresh){
+                        for(unsigned long k=0;k<(*samples)[i].shapeResiduals.size();k++){
+                            (*leftsumresiduals)[j][k]=(*leftsumresiduals)[j][k]+(*samples)[i].shapeResiduals[k];
+                        }
+                    }
+                }
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+        vector< vector<Point2f> >* leftsumresiduals;
+        vector<unsigned long>* left_count;
+        unsigned long* num_test_splits;
+        vector<splitr>* feats;
+};
+
+splitr FacemarkKazemiImpl::getTestSplits(vector<Point2f> pixel_coordinates,int seed)
+{
+    splitr feat;
+    //generates splits whose probability is above a particular threshold.
+    //P(u,v)=e^(-distance/lambda) as described in the research paper
+    //cited above. This helps to select closer pixels hence make efficient
+    //splits.
```

Comment:
Sir according to the paper the values of both threshold and probability are being randomly generated. So are we trying to provide users a functionality so that they can set their own threshold value or the way these values are generated ?